WIP: EpochX step 1 - basic infrastructure

valassi opened the pull request:
This is step1 in the plan outlined in issue #244.

It just includes a script to automate the code generation and the comparison to a previous version. I used Olivier's tutorial plugin as-is (with some renaming) and I generated eemumu.

The problem is that _code generation is not stable_. Repeating the same identical code generation twice produces code which is functionally identical, but is NOT line-by-line identical. The order in which some functions are generated (or even the order in which some lines within the same functions are generated) is not strictly reproducible and may vary from one run to another. @oliviermattelaer do you have any feedback on this? I will investigate anyway.

Example,
```
...
Files ee_mumu.NEW/src/mgOnGpuConfig.h and ee_mumu/src/mgOnGpuConfig.h are identical
Files ee_mumu.NEW/src/mgOnGpuTypes.h and ee_mumu/src/mgOnGpuTypes.h are identical
diff -rs ee_mumu.NEW/src/Parameters_sm.cc ee_mumu/src/Parameters_sm.cc
46d45
<   mdl_CKM3x3 = 1.; 
48a48
>   mdl_CKM3x3 = 1.; 
132,133d131
<   cout << setw(20) <<  "mdl_CKM3x3 " <<  "= " << setiosflags(ios::scientific)
<       << setw(10) << mdl_CKM3x3 << endl;
137a136,137
>   cout << setw(20) <<  "mdl_CKM3x3 " <<  "= " << setiosflags(ios::scientific)
>       << setw(10) << mdl_CKM3x3 << endl;
diff -rs ee_mumu.NEW/src/Parameters_sm.h ee_mumu/src/Parameters_sm.h
26,31c26,32
<         mdl_Gf, aEWM1, mdl_MH, mdl_MZ, mdl_MTA, mdl_MT, mdl_MB, mdl_CKM3x3,
<         mdl_conjg__CKM1x1, mdl_conjg__CKM3x3, mdl_MZ__exp__2, mdl_MZ__exp__4,
<         mdl_sqrt__2, mdl_MH__exp__2, mdl_aEW, mdl_MW, mdl_sqrt__aEW, mdl_ee,
<         mdl_MW__exp__2, mdl_sw2, mdl_cw, mdl_sqrt__sw2, mdl_sw, mdl_g1, mdl_gw,
<         mdl_vev, mdl_vev__exp__2, mdl_lam, mdl_yb, mdl_yt, mdl_ytau, mdl_muH,
<         mdl_ee__exp__2, mdl_sw__exp__2, mdl_cw__exp__2;
---
>         mdl_Gf, aEWM1, mdl_MH, mdl_MZ, mdl_MTA, mdl_MT, mdl_MB,
>         mdl_conjg__CKM1x1, mdl_conjg__CKM3x3, mdl_CKM3x3, mdl_MZ__exp__2,
>         mdl_MZ__exp__4, mdl_sqrt__2, mdl_MH__exp__2, mdl_aEW, mdl_MW,
>         mdl_sqrt__aEW, mdl_ee, mdl_MW__exp__2, mdl_sw2, mdl_cw, mdl_sqrt__sw2,
>         mdl_sw, mdl_g1, mdl_gw, mdl_vev, mdl_vev__exp__2, mdl_lam, mdl_yb,
>         mdl_yt, mdl_ytau, mdl_muH, mdl_ee__exp__2, mdl_sw__exp__2,
>         mdl_cw__exp__2;
Files ee_mumu.NEW/src/rambo.cc and ee_mumu/src/rambo.cc are identical
...
```

valassi commented:
The issue which may be the easiest is this, and it is printed out in the log
```
< ALOHA: aloha creates routines (starting by FFV2)
---
> ALOHA: aloha creates routines (starting by FFV1)
```

roiser commented:
Hi @valassi, thanks for going ahead quickly with this work. Please don't merge this yet, as I would still have some comments on the development model in #244 but no time yet to put them all together in a reply. thanks

valassi commented:
Good old debug printouts, from `logger.debug("UFOModelConverterCPP wanted_lorentz : %s" % str(wanted_lorentz) )` in the UFOModelConverterCPP __init__ in madgraph/iolibs/export_cpp.py
```
< DEBUG: UFOModelConverterCPP wanted_lorentz : [(('FFV2', 'FFV4'), (), 3), (('FFV2', 'FFV4'), (), 0), (('FFV1',), (), 0), (('FFV1',), ('P0',), 3)] 
---
> DEBUG: UFOModelConverterCPP wanted_lorentz : [(('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 3)] 
```

Note that I am using Python3.8, so in principle standard dictionaries {} are now ordered, in the sense that they respect the order in which items were added. So I am now checking where some lists and dictionaries are created...

valassi commented:
A few more debug printouts. It looks as if the 'randomness' (non reproducible behavious) is introduced in HelasMultiProcess?

This is a diff of two subsequent runs, with different outputs
```
madgraph_interface 8348 _curr_matrix_elements : <class 'madgraph.iolibs.group_subprocs.SubProcessGroupList'>
INFO: group_subprocs 520 self : <class 'madgraph.iolibs.group_subprocs.SubProcessGroupList'> 
INFO: helas_objects 5006 self : <class 'madgraph.core.helas_objects.HelasMatrixElement'> 
INFO: helas_objects 5006 output : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)] 
madgraph_interface 8349 wanted_lorentz : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)]
madgraph_interface 8371 wanted_lorentz : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)]
[1;32mDEBUG:  create the model [1;30m[output.py at line 92][0m [0m
PLUGIN.CUDACPP_SA_OUTPUT.output: convert_model 90 wanted_lorentz : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)][0m
[1;32mDEBUG: convert_model 2505 wanted_lorentz : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)] [0m
[1;32mDEBUG: UFOModelConverterGPU wanted_lorentz : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)] [0m
[1;32mDEBUG: UFOModelConverterCPP wanted_lorentz : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)] [0m
---
madgraph_interface 8348 _curr_matrix_elements : <class 'madgraph.iolibs.group_subprocs.SubProcessGroupList'>
INFO: group_subprocs 520 self : <class 'madgraph.iolibs.group_subprocs.SubProcessGroupList'> 
INFO: helas_objects 5006 self : <class 'madgraph.core.helas_objects.HelasMatrixElement'> 
INFO: helas_objects 5006 output : [(('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3), (('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0)] 
madgraph_interface 8349 wanted_lorentz : [(('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3)]
madgraph_interface 8371 wanted_lorentz : [(('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3)]
[1;32mDEBUG:  create the model [1;30m[output.py at line 92][0m [0m
PLUGIN.CUDACPP_SA_OUTPUT.output: convert_model 90 wanted_lorentz : [(('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3)][0m
[1;32mDEBUG: convert_model 2505 wanted_lorentz : [(('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3)] [0m
[1;32mDEBUG: UFOModelConverterGPU wanted_lorentz : [(('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3)] [0m
[1;32mDEBUG: UFOModelConverterCPP wanted_lorentz : [(('FFV1',), (), 0), (('FFV2', 'FFV4'), (), 0), (('FFV1',), ('P0',), 3), (('FFV2', 'FFV4'), (), 3)] [0m
```

oliviermattelaer commented:
Yes, we quite often use code like this to remove duplicate
mylist = list(set(mylist))
This works as expected but can indeed lead to --physically irrelevant -- randomness within the generated code.

The particular issue under consideration might be linked to the following line:
./helas_objects.py:5617:        return list(set(helas_list))
./helas_objects.py:5628:        return list(set(coupling_list))

I guess I could use this: list(dict.fromkeys(items))
to preserve the ordering but this will obviously slow down the generation code

valassi commented:
Ok the bug (or at least the first one) is most likely here?
```
    def get_used_lorentz(self):
        """Return a list of (lorentz_name, conjugate, outgoing) with
        all lorentz structures used by this HelasMultiProcess."""
        helas_list = []

        for me in self.get('matrix_elements'):
            helas_list.extend(me.get_used_lorentz())
                
	return list(set(helas_list))
```

While helas_list is reproducible, passing it through list(set(helas_list)) makes it random... what was the purpose of that?!

valassi commented:
Ups I had actuallycreated that before but I had not sent the comment with enter...
Thanks Olivier for confirming that indeed this is the right place!

Now I created a patch and a bug report in bazaar
https://bugs.launchpad.net/mg5amcnlo/+bug/1944444

Anyway, I created also the patch in this epochX WIP MR
https://github.com/madgraph5/madgraph4gpu/pull/245/commits/d4e07dd6415ce64a8146fad985ad155153202053
I copy the files to the MG installation.

So I do not need this fixed upstream yet.
Indeed you ar eright that speed may be an issue for larger processes... I would still suggest that we keep the generated code reproducible though.

Note that the solution I used is https://stackoverflow.com/a/17016257
Here the author says that it is now the fastest using the current python3 dictionaries, so maybe speed will no be not such an issue

valassi commented:
Thanks to Olivier, the resproducibility issue is now fixed. I have produced basic code for eemuu and ggttgg. In th eplugin I only added one minimal change to the Makefile, to bypass some build issue for tests, that I will fix more carefully in later versions of the c++/cuda code from epoch1/2. I added a basic throughput script and ran some profiling, logs are in the repo. Probably I also need to switch on fast math to reproduce the results I got in epoch1/2.

This completes my 'step1' PR #245 in the plan outlined in issue #244. Still a long way to go...

I keep this in WIP even if essentially complete. I will open a second PR (including this one) for the next steps.
